runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java

```diff
@@ -341,7 +342,16 @@ private void processTrimmedException(TrimmedException te) {
 
         // The list to store read results in
         List<ILogData> readFrom = readAll(toRead).stream()
-                .filter(x -> x.getType() == DataType.DATA)
+                // If the last address is a hole, we want to update the
+                // global pointer to include that address as well.
+                // During a checkpoint, a hole is appended at the end of each
+                // stream which might end up being used as a trim mark in the
+                // next trim cycle. If there are no updates between the checkpoint
+                // and the trim cycle for that stream, there is a chance that the
+                // last address (hole) will be trimmed, causing the TrimmedException
+                // to be thrown. By updating the global pointer to include the hole,
+                // we prevent that from happening.
+                .filter(x -> x.isData() || x.isHole())
```

Comment:
Isn't this what @zhangn49 is addressing in PR #2387?

Reply:
If you try running the unit tests from this patch with #2387 changes applies minus this line, you will see that the tests are still failing. I can go into more detail, but it in some circumstances we do not set the global pointer within `remainingUpTo()`:

```
        // Nothing read, nothing to process.
        if (entries.size() == 0) {
            // We've resolved up to maxGlobal, so remember it. (if it wasn't max)
            if (maxGlobal != Address.MAX) {
                // Set Global Pointer and check that it is not pointing to an address in the trimmed space.
                getCurrentContext().setGlobalPointerCheckGCTrimMark(maxGlobal);
            }
            return entries;
        }
```

And:
```
        // Otherwise update the pointer
        if (maxGlobal != Address.MAX) {
            // Set Global Pointer and check that it is not pointing to an address in the trimmed space.
            getCurrentContext().setGlobalPointerCheckGCTrimMark(maxGlobal);
        } else {
            // Update pointer from log data and then validate final position of the pointer against GC trim mark.
            updatePointer(entries.get(entries.size() - 1));
            getCurrentContext().validateGlobalPointerPosition(getCurrentGlobalPosition());
        }
```

Which means we do not end up exercising changes within  #2387. In one of my other comments I mention a need for a clean up -- this is one of the things I was referring to.